syslog-ng-ctl/syslog-ng-ctl.c

```diff
@@ -347,6 +377,127 @@ static GOptionEntry no_options[] =
   { NULL, 0, 0, G_OPTION_ARG_NONE, NULL, NULL }
 };
 
+static void
+set_console_echo(gboolean new_state)
+{
+  struct termios t;
+
+  if (tcgetattr(STDIN_FILENO, &t))
+    {
+      fprintf(stderr, "syslog-ng-ctl: error while tcgetattr: %s\n", strerror(errno));
+      return;
+    }
+
+  if (new_state)
+    t.c_lflag |= ECHO;
+  else
+    t.c_lflag &= ~((tcflag_t) ECHO);
+
+  if (tcsetattr(STDIN_FILENO, TCSANOW, &t))
+    fprintf(stderr, "syslog-ng-ctl: error while tcsetattr: %s\n", strerror(errno));
+}
+
+static void
+get_password_from_stdin(gchar *buffer, gulong *length)
+{
+  printf("enter password:");
+  set_console_echo(FALSE);
+  if (-1 == getline(&buffer, length, stdin))
+    {
+      set_console_echo(TRUE);
+      fprintf(stderr, "error while reading password from terminal: %s", strerror(errno));
+      g_assert_not_reached();
+    }
+  set_console_echo(TRUE);
+  printf("\n");
+}
+
+static gboolean
+is_syslog_ng_running()
+{
+  return control_client_connect(control_client);
+}
+
+static gchar *
+fetch_next_remaining(gchar **remaining, gint *available_index)
+{
+  if (!remaining)
+    return NULL;
+
+  return remaining[(*available_index)++];
+}
+
+static gint
+slng_passwd_add(int argc, char *argv[], const gchar *mode, GOptionContext *ctx)
+{
+  gchar *answer;
+  gint remaining_unused_index = 0;
+
+
+  if (!credentials_key)
+    credentials_key = fetch_next_remaining(credentials_remaining, &remaining_unused_index);
+
+  if (!credentials_key)
+    {
+      gchar *usage = g_option_context_get_help(ctx, TRUE, NULL);
+      fprintf(stderr, "Error: missing arguments!\n%s\n", usage);
+      g_free(usage);
+      return 1;
+    }
+
+  if (!is_syslog_ng_running())
+    return 1;
+
+  if (!credentials_secret)
+    credentials_secret = fetch_next_remaining(credentials_remaining, &remaining_unused_index);
+
+  gchar *secret_to_store;
+  if (credentials_secret)
+    {
+      secret_to_store = g_strdup(credentials_secret);
+      if (!secret_to_store)
+        g_assert_not_reached();
+    }
+  else
+    {
+      gulong buff_size = 256;
```

Comment:
I'd use `gsize` instead of `gulong`.

If this size limitation is intentional, we should document it.
I'd use a dynamically sized buffer here. **Reading passwords from the command line (as CL parameters) has no such limitation, so I think it's weird to restrict the stdin version (which will more likely be used in production).**